base/utf.jl

```diff
@@ -0,0 +1,880 @@
+
+    function UTF32String(data::Vector{Char})
+        if length(data) < 1 || data[end] != Char(0)
+            utf_errfunc(UTF_ERR_NULL_32_TERMINATE, 0, 0)
+        end
+        new(data)
+    end
+end
+UTF32String(data::Vector{UInt32}) = UTF32String(reinterpret(Char, data))
+
+is_surrogate_lead(c::Unsigned) = ((c & ~0x003ff) == 0xd800)
+is_surrogate_trail(c::Unsigned) = ((c & ~0x003ff) == 0xdc00)
+is_surrogate_char(c::Unsigned) = ((c & ~0x007ff) == 0xd800)
+is_valid_continuation(c) = ((c & 0xc0) == 0x80)
+
+function length(s::UTF16String)
+    d = s.data
+    len = length(d) - 1
+    len == 0 && return 0
+    cnum = 0
+    for i = 1:len
+        @inbounds cnum += !is_surrogate_trail(d[i])
+    end
+    cnum
+end
+
+function endof(s::UTF16String)
+    d = s.data
+    i = length(d) - 1
+    i == 0 && return i
+    return is_surrogate_char(d[i]) ? i-1 : i
+end
+
+get_supplementary(lead::Unsigned, trail::Unsigned) = (UInt32(lead-0xd7f7)<<10 + trail)
+
+function next(s::UTF16String, i::Int)
+    ch = s.data[i]
+    !is_surrogate_char(ch) && return (Char(ch), i+1)
+    # check length, account for terminating \0
+    i >= (length(s.data)-1) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, i, UInt32(ch))
+    !is_surrogate_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, i, ch)
+    ct = s.data[i+1]
+    !is_surrogate_trail(ct) && utf_errfunc(UTF_ERR_NOT_TRAIL, i, ch)
+    Char(get_supplementary(ch, ct)), i+2
+end
+
+function reverseind(s::UTF16String, i::Integer)
+    j = length(s.data) - i
+    return Base.is_surrogate_trail(s.data[j]) ? j-1 : j
+end
+
+lastidx(s::UTF16String) = length(s.data) - 1 # s.data includes NULL terminator
+
+function reverse(s::UTF16String)
+    d = s.data
+    out = similar(d)
+    out[end] = 0 # NULL termination
+    n = length(d)
+    @inbounds for i = 1:n-1
+        ch = d[n-i]
+        if Base.is_surrogate_lead(ch)
+            out[i],out[i-1] = out[i-1],ch
+        else
+            out[i] = ch
+        end
+    end
+    UTF16String(out)
+end
+
+next(s::UTF32String, i::Int) = (s.data[i], i+1)
+endof(s::UTF32String) = length(s.data) - 1
+length(s::UTF32String) = length(s.data) - 1
+
+const UTF_NO_LONG_NULL = 1      # don't accept 0xc0 0x80 for '\0'
+const UTF_NO_SURROGATES = 2     # don't accept surrogate pairs in UTF-8/UTF-32
+const UTF_ACCEPT_LONG = 4       # accept long encodings (other than long null in UTF-8)
+
+const UTF_LONG = 1              # Long encodings are present
+const UTF_LATIN1 = 2            # characters in range 0x80-0xFF present
+const UTF_UNICODE2 = 4          # characters in range 0x100-0x7ff present
+const UTF_UNICODE3 = 8          # characters in range 0x800-0xd7ff, 0xe000-0xffff
+const UTF_UNICODE4 = 16         # non-BMP characters present
+const UTF_SURROGATE = 32        # surrogate pairs present
+
+# Get a UTF-8 continuation byte, give error if invalid, and update position and character value
+macro get_continuation!(ch, byt, str, pos)
```

Comment:
Why make it mutate its first argument? Since it's a scalar, there won't be any allocation anyway. Better make it return the value.

Also, why should this be a macro? It's a standard function AFAICT.